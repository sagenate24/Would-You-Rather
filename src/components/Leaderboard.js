import React, { Component } from 'react';
import { connect } from 'react-redux';
import '../styles/Leaderboard.css';

class Leaderboard extends Component {
  organizeLeaderBoard(id) {
    if (id === this.props.userDetails[0].id) {
      return '1';
    } else if (id === this.props.userDetails[1].id) {
      return '2';
    } else if (id === this.props.userDetails[2].id) {
      return '3';
    } else { return null; }
  }
  
  render() {
    const { userDetails } = this.props;
    return (
      <div>
        <h1>Leaderboard</h1>
        {userDetails.map((user) => {
          return (
            <div key={user.id} className='leaderboard'>
              <img src={user.avatarURL} alt='avatar' className='avatar' />
              <div className='leaderboard_description'>
                <h3>{user.name}</h3>
                <p> questions asked: {user.questionsAsked}</p>
                <p>questions answered: {user.questionsAnswered}</p>
                <p>RANK: {this.organizeLeaderBoard(user.id)}</p>
              </div>
            </div>
          );
        })}
      </div>
    );
  }
}

function mapStateToProps({ users }) {
  //create a new array of objects for each user created
  const userDetails = Object.keys(users).map((user) => {
    const userObject = {
      id: users[user].id,
      avatarURL: users[user].avatarURL,
      name: users[user].name,
      questionsAsked: users[user].questions.length,
      questionsAnswered: Object.keys(users[user].answers).length,
    }
    const total = userObject.questionsAsked + userObject.questionsAnswered;
    userObject.total = total;
    return (userObject);

  }).sort((a, b) => (
    b - a
  ));

  return {
    userDetails,
  }
}

export default connect(mapStateToProps)(Leaderboard);
