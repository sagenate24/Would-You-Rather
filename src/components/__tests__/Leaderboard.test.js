import * as NathanUrl from '../../Images/me.png';
import * as ClaireUrl from '../../Images/claire.jpg';
import * as JohnDoe from '../../Images/johnDoe.png';
import * as trophie from '../../Images/trophie.png';
import * as trophie2 from '../../Images/trophie2.png';
import * as trophie3 from '../../Images/trophie3.png';
import * as renderer from 'react-test-renderer';

import React from 'react';
import { shallow, mount } from 'enzyme';
import Leaderboard from '../Leaderboard';

const mockUsers = [
  {
    id: "nathan",
    name: "Nathan Sage",
    avatarURL: NathanUrl,
    questionsAnswered: 4,
    questionsAsked: 4,
    total: 8,
  },
  {
    id: "claire",
    name: "Claire Teters",
    avatarURL: ClaireUrl,
    questionsAnswered: 4,
    questionsAsked: 3,
    total: 7,
  },
  {
    id: "johndoe",
    name: "John Doe",
    avatarURL: JohnDoe,
    questionsAnswered: 4,
    questionsAsked: 2,
    total: 6,
  }
];

it('should render a list  users', () => {

  const tree = renderer.create(
    <Leaderboard.WrappedComponent users={mockUsers} />
  ).toJSON();
  expect(tree).toMatchSnapshot();

});

it('should not blow up into smithereens if there are no users', () => {

  const tree = renderer.create(
    <Leaderboard.WrappedComponent />
  ).toJSON();
  expect(tree).toMatchSnapshot();

});

it('should render a list of only one user', () => {

  const mockUser = [
    {
      id: "nathan",
      name: "Nathan Sage",
      avatarURL: NathanUrl,
      questionsAnswered: 4,
      questionsAsked: 4,
      total: 8,
    }
  ];

  const wrapper = mount(<Leaderboard.WrappedComponent users={mockUser} />);

  expect(wrapper.find('div.user-container').children()).toHaveLength(1);

});

it('should return a trophie if the given id exsists', () => {

  const wrapper = shallow(<Leaderboard.WrappedComponent users={mockUsers} />);

  expect(wrapper.instance().chooseTrophey('nathan')).toEqual(trophie);
  expect(wrapper.instance().chooseTrophey('claire')).toEqual(trophie2);
  expect(wrapper.instance().chooseTrophey('johndoe')).toEqual(trophie3);
  expect(wrapper.instance().chooseTrophey()).toEqual(null);

});
