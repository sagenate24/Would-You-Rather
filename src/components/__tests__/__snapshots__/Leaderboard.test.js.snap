// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`should not blow up into smithereens if there are no users 1`] = `
<div>
  <div
    style={
      Object {
        "opacity": 0,
        "position": "relative",
        "top": 20,
        "transition": "opacity 400ms, top 400ms",
      }
    }
  >
    <div
      className="leader-container"
    >
      <div
        className="container-header"
      >
        <p
          className="leader-text-header"
        >
          Leaderboard 
          <span
            aria-label="Trophie"
            role="img"
            style={
              Object {
                "paddingLeft": "5px",
              }
            }
          >
            üèÜ
          </span>
        </p>
      </div>
      <div
        className="user-container"
      />
    </div>
  </div>
</div>
`;

exports[`should render a list  users 1`] = `
<div>
  <div
    style={
      Object {
        "opacity": 0,
        "position": "relative",
        "top": 20,
        "transition": "opacity 400ms, top 400ms",
      }
    }
  >
    <div
      className="leader-container"
    >
      <div
        className="container-header"
      >
        <p
          className="leader-text-header"
        >
          Leaderboard 
          <span
            aria-label="Trophie"
            role="img"
            style={
              Object {
                "paddingLeft": "5px",
              }
            }
          >
            üèÜ
          </span>
        </p>
      </div>
      <div
        className="user-container"
      >
        <div
          className="leaderboard-body"
          data-id="nathan"
        >
          <img
            alt="rank"
            className="leader-trophey"
            src={
              Object {
                "0": "t",
                "1": "r",
                "10": "g",
                "2": "o",
                "3": "p",
                "4": "h",
                "5": "i",
                "6": "e",
                "7": ".",
                "8": "p",
                "9": "n",
                "default": "trophie.png",
              }
            }
          />
          <div
            className="leader-avatar"
          >
            <img
              alt="avatar"
              className="avatar-leaderboard"
              src={
                Object {
                  "0": "m",
                  "1": "e",
                  "2": ".",
                  "3": "p",
                  "4": "n",
                  "5": "g",
                  "default": "me.png",
                }
              }
            />
          </div>
          <div
            className="leader-info"
          >
            <h3>
              Nathan Sage
            </h3>
            <p>
               questions asked: 
              <span>
                4
              </span>
            </p>
            <p>
              questions answered: 
              <span>
                4
              </span>
            </p>
          </div>
          <div
            className="leader-score"
          >
            <div
              className="leader-score-header"
            >
              <p>
                Score
              </p>
            </div>
            <span
              className="leader-total-number"
            >
              8
            </span>
          </div>
        </div>
        <div
          className="leaderboard-body"
          data-id="claire"
        >
          <img
            alt="rank"
            className="leader-trophey"
            src={
              Object {
                "0": "t",
                "1": "r",
                "10": "n",
                "11": "g",
                "2": "o",
                "3": "p",
                "4": "h",
                "5": "i",
                "6": "e",
                "7": "2",
                "8": ".",
                "9": "p",
                "default": "trophie2.png",
              }
            }
          />
          <div
            className="leader-avatar"
          >
            <img
              alt="avatar"
              className="avatar-leaderboard"
              src={
                Object {
                  "0": "c",
                  "1": "l",
                  "2": "a",
                  "3": "i",
                  "4": "r",
                  "5": "e",
                  "6": ".",
                  "7": "j",
                  "8": "p",
                  "9": "g",
                  "default": "claire.jpg",
                }
              }
            />
          </div>
          <div
            className="leader-info"
          >
            <h3>
              Claire Teters
            </h3>
            <p>
               questions asked: 
              <span>
                3
              </span>
            </p>
            <p>
              questions answered: 
              <span>
                4
              </span>
            </p>
          </div>
          <div
            className="leader-score"
          >
            <div
              className="leader-score-header"
            >
              <p>
                Score
              </p>
            </div>
            <span
              className="leader-total-number"
            >
              7
            </span>
          </div>
        </div>
        <div
          className="leaderboard-body"
          data-id="johndoe"
        >
          <img
            alt="rank"
            className="leader-trophey"
            src={
              Object {
                "0": "t",
                "1": "r",
                "10": "n",
                "11": "g",
                "2": "o",
                "3": "p",
                "4": "h",
                "5": "i",
                "6": "e",
                "7": "3",
                "8": ".",
                "9": "p",
                "default": "trophie3.png",
              }
            }
          />
          <div
            className="leader-avatar"
          >
            <img
              alt="avatar"
              className="avatar-leaderboard"
              src={
                Object {
                  "0": "j",
                  "1": "o",
                  "10": "g",
                  "2": "h",
                  "3": "n",
                  "4": "D",
                  "5": "o",
                  "6": "e",
                  "7": ".",
                  "8": "p",
                  "9": "n",
                  "default": "johnDoe.png",
                }
              }
            />
          </div>
          <div
            className="leader-info"
          >
            <h3>
              John Doe
            </h3>
            <p>
               questions asked: 
              <span>
                2
              </span>
            </p>
            <p>
              questions answered: 
              <span>
                4
              </span>
            </p>
          </div>
          <div
            className="leader-score"
          >
            <div
              className="leader-score-header"
            >
              <p>
                Score
              </p>
            </div>
            <span
              className="leader-total-number"
            >
              6
            </span>
          </div>
        </div>
      </div>
    </div>
  </div>
</div>
`;
